plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/CachingDirectoryLister.java

```diff
@@ -145,32 +162,37 @@ public void invalidate(Partition partition)
         delegate.invalidate(partition);
     }
 
-    private RemoteIterator<LocatedFileStatus> cachingRemoteIterator(ValueHolder cachedValueHolder, RemoteIterator<LocatedFileStatus> iterator, Path path)
+    private RemoteIterator<LocatedFileStatus> cachingRemoteIterator(ValueHolder cachedValueHolder, Path path)
     {
         return new RemoteIterator<>()
         {
-            private final List<LocatedFileStatus> files = new ArrayList<>();
+            private int fileIndex;
 
             @Override
             public boolean hasNext()
                     throws IOException
             {
-                boolean hasNext = iterator.hasNext();
-                if (!hasNext) {
+                try {
+                    boolean hasNext = cachedValueHolder.hasNext(fileIndex);
+                    // Update cache weight of cachedValueHolder for a given path.
                     // The cachedValueHolder acts as an invalidation guard. If a cache invalidation happens while this iterator goes over
                     // the files from the specified path, the eventually outdated file listing will not be added anymore to the cache.
-                    cache.asMap().replace(path, cachedValueHolder, new ValueHolder(files));
+                    cache.asMap().replace(path, cachedValueHolder, cachedValueHolder);
+                    return hasNext;
+                }
+                catch (IOException exception) {
+                    // invalidate cached value to force retry of directory listing
+                    cache.invalidate(path);
+                    throw exception;
```

Comment:
This prevents persistent failure. We still have exception sharing though, which is undesirable. see eg https://github.com/trinodb/trino/issues/11067 

this can eg  lead to spurious query failures when using extra credentials to access GCS

Reply:
> Only concurrent listings will throw same exception, but that is desirable because we don't want to scan partially listed table.

That's desirable when concurrent scans are from same user, same session.
Which is guaranteed for per-transaction case, but is not satisfied for global cache.

> TBH I'm not sure non-transactional caching DirectoryLister works with extra credentials.

What do you mean?
Do we prevent it from being used when extra credentials are in use?
Or, when eg HDFS impersonation is used?

Reply:
> That's desirable when concurrent scans are from same user, same session.

Session doesn't affect listing.
If we use same cache entries within global cache for different impersonated users, that's probably a bug, but it's out of scope of the PR

> Do we prevent it from being used when extra credentials are in use?
> Or, when eg HDFS impersonation is used?

I don't think we prevent global cache in such cases. Probably we should have cache key that takes impersonated user into account, but that's outside of scope of this PR.

Reply:
> Session doesn't affect listing.

it does (at least for "result vs failure"), when you have eg HDFS impersonation, or object storage extra credentials (this is what the extra credentials were originally added for).

> If we use same cache entries within global cache for different impersonated users, that's probably a bug, but it's out of scope of the PR

this PR adds exception sharing. I don't think it was there before the change, so it seems to me it's in scope.